### PR TITLE
Validation of number of in- and outneighbors by node type

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Fixes https://github.com/Deltares/Ribasim/issues/333.

### CODE REVIEW

**Comment by SouthEndMusic** [comment]
For the model in the issue you now get this:

```
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #5).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #7).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #9).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #11).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #13).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #15).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #17).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #19).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #21).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #23).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #25).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #27).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #29).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #31).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #33).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #35).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #37).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #39).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #41).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #43).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #45).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #47).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #49).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #51).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #53).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #55).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #57).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #59).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #61).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #63).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #65).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #67).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #69).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #71).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #73).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #75).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #77).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #79).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #81).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #83).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #85).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #87).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #89).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #91).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #93).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #95).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #97).
│ Nodes of type Ribasim.ManningResistance must have at least 1 inneighbors (got 0 for node #99).
```

File: core/src/bmi.jl [hunk 1]
```diff
@@ -16,6 +16,10 @@ function BMI.initialize(T::Type{Model}, config::Config)::Model
     local parameters, state, n
     try
         parameters = Parameters(db, config)
+        if !valid_flow_neighbor_amounts(parameters)
+            error("Invalid neighbor amounts.")
```
Comment (Huite) [inline core/src/bmi.jl:20]: I think "amounts" isn't the appropriate when we're talking about the number of neighbors. (Amount generally refers to uncountable entities.)

This is also a little terse (although I guess the error message should make enough sense will everything that has been logged via @error).

I would rephrase to something like "Invalid network definition:" so it acts as a header and then the content of errors will clarify it.

File: core/src/solve.jl [hunk 2]
```diff
@@ -295,6 +299,93 @@ struct Parameters
     lookup::Dict{Int, Symbol}
 end
 
+"""
+Test for each node given its node type whether it has an allowed
+number of flow inneighbors and flow outneighbors
+"""
+function valid_flow_neighbor_amounts(p::Parameters)::Bool
```
Comment (Huite) [inline core/src/solve.jl:306]: `valid_flow_neighbor_numbers` or `valid_flow_n_neighbor` (or something?)

File: core/src/solve.jl [hunk 3]
```diff
@@ -295,6 +299,93 @@ struct Parameters
...
+    append!(errors, valid_flow_neighbor_amounts(graph_flow, linear_resistance))
+    append!(errors, valid_flow_neighbor_amounts(graph_flow, manning_resistance))
+    append!(errors, valid_flow_neighbor_amounts(graph_flow, tabulated_rating_curve))
+    append!(errors, valid_flow_neighbor_amounts(graph_flow, fractional_flow))
+    append!(errors, valid_flow_neighbor_amounts(graph_flow, level_boundary))
+    append!(errors, valid_flow_neighbor_amounts(graph_flow, flow_boundary))
+    append!(errors, valid_flow_neighbor_amounts(graph_flow, pump))
+    append!(errors, valid_flow_neighbor_amounts(graph_flow, terminal))
+
+    return if isempty(errors)
```
Comment (Huite) [inline core/src/solve.jl:334]: Now, I don't really like implicit returns, and I guess this is explicit.... but I (personally) still don't like it.

I guess either do:

```julia
if isempty(errors)
    true
else
    @error join(errors, "\n")
    false
end
```

Or just nice and explicit as I prefer it ^^:

```julia
if isempty(errors)
    return true
else
    @error join(errors, "\n")
    return false
end
```

File: core/src/solve.jl [hunk 4]
```diff
@@ -295,6 +299,93 @@ struct Parameters
...
+        true
+    else
+        @error join(errors, "\n")
+        false
+    end
+end
+
+function valid_flow_neighbor_amounts(
```
Comment (Huite) [inline core/src/solve.jl:342]: Also n_neighbor

File: core/test/validation.jl [hunk 5]
```diff
@@ -0,0 +1,23 @@
+using Ribasim
+using Graphs: DiGraph, add_edge!
+
+@testset "Neighbor count validation" begin
```
Comment (Huite) [inline core/test/validation.jl:4]: While we're at it, we should probably also add a test for all (four) messages.

Since the errors are nicely concatenated, this can be done in a single test.

**Review by Huite — DISMISSED** [review]

File: core/src/validation.jl [hunk 6]
```diff
@@ -56,12 +56,44 @@ neighbortypes(::Val{:FlowBoundary}) =
...
 
+# Allowed number of inneighbors and outneighbors per node type
+struct neighbor_amounts_bounds
+    in_min::Int
+    in_max::Int
+    out_min::Int
+    out_max::Int
+end
+
+neighbors_amounts_bounds(nodetype::Symbol) = neighbors_amounts_bounds(Val(nodetype))
```
Comment (Huite) [inline core/src/validation.jl:79]: Can you also replace these mentions of "amounts" by number or n?

**Review by Huite — APPROVED** [review]
